### PR TITLE
fix #take/#drop functions

### PR DESCRIPTION
**Pull request by sskeirik**
Fix by doing an explicit case analysis on the subtype injected into
`Val` which is either `IVal` or `FVal`.

### CODE REVIEW

File: data.md [hunk 1]
```diff
@@ -264,6 +267,16 @@ We also add `undefined` as a value, which makes many partial functions in the se
  // --------------------------
 ```
 
+The `getXXXXType` takes a `Val` and projects out its `IValType` or `FValType`.
+
+```k
+    syntax IValType ::= getIValType ( IVal ) [function]
+    syntax FValType ::= getFValType ( FVal ) [function]
+ // -----------------------------------------------
```
Comment (ehildenb) [inline data.md]:
```suggestion
 // ---------------------------------------------------
```

File: data.md [hunk 2]
```diff
@@ -264,6 +267,16 @@ We also add `undefined` as a value, which makes many partial functions in the se
...
+    rule getIValType( < ITYPE > N) => ITYPE
+    rule getFValType( < FTYPE > F) => FTYPE
```
Comment (ehildenb) [inline data.md]:
```suggestion
    rule getIValType( < ITYPE > N ) => ITYPE
    rule getFValType( < FTYPE > F ) => FTYPE
```

File: data.md [hunk 3]
```diff
@@ -358,11 +371,17 @@ One needs to unname the `ValTypes` first before calling the `#take` or `#drop` f
...
-    rule #take(.ValTypes,   _)                              => .ValStack
-    rule #take(TYPE:AValType VTYPES, < TYPE > VAL:Number : VALSTACK) => < TYPE > VAL : #take(VTYPES, VALSTACK)
-
-    rule #drop(.ValTypes,   VALSTACK)                       => VALSTACK
-    rule #drop(TYPE:AValType VTYPES, < TYPE > VAL:Number : VALSTACK) => #drop(VTYPES, VALSTACK)
+    rule #take(.ValTypes,   _             ) => .ValStack
+    rule #take(TYPE VTYPES, VAL : VALSTACK) => VAL : #take(VTYPES, VALSTACK)
+      requires unnameValType(TYPE) ==K getIValType(project:IVal(VAL))
+    rule #take(TYPE VTYPES, VAL : VALSTACK) => VAL : #take(VTYPES, VALSTACK)
+      requires unnameValType(TYPE) ==K getFValType(project:FVal(VAL))
```
Comment (ehildenb) [inline data.md]:
```suggestion
    rule #take(TYPE VTYPES, VAL : VALSTACK) => VAL : #take(VTYPES, VALSTACK) requires unnameValType(TYPE) ==K getIValType(project:IVal(VAL))
    rule #take(TYPE VTYPES, VAL : VALSTACK) => VAL : #take(VTYPES, VALSTACK) requires unnameValType(TYPE) ==K getFValType(project:FVal(VAL))
```

File: data.md [hunk 4]
```diff
@@ -358,11 +371,17 @@ One needs to unname the `ValTypes` first before calling the `#take` or `#drop` f
...
+
+    rule #drop(.ValTypes,   VALSTACK      ) => VALSTACK
+    rule #drop(TYPE VTYPES, VAL : VALSTACK) => #drop(VTYPES, VALSTACK)
+      requires unnameValType(TYPE) ==K getIValType(project:IVal(VAL))
+    rule #drop(TYPE VTYPES, VAL : VALSTACK) => #drop(VTYPES, VALSTACK)
+      requires unnameValType(TYPE) ==K getFValType(project:FVal(VAL))
```
Comment (ehildenb) [inline data.md]:
```suggestion
    rule #drop(TYPE VTYPES, VAL : VALSTACK) => #drop(VTYPES, VALSTACK) requires unnameValType(TYPE) ==K getIValType(project:IVal(VAL))
    rule #drop(TYPE VTYPES, VAL : VALSTACK) => #drop(VTYPES, VALSTACK) requires unnameValType(TYPE) ==K getFValType(project:FVal(VAL))
```

**Comment by sskeirik** [comment]
The merge is failing because the jenkins build is failing. Do you know anything about how to fix that?

**Comment by hjorthjort** [comment]
Is it working on your machine?

Stephen Skeirik <notifications@github.com> schrieb am Di., 10. März 2020,
7:01 PM:

**Comment by ehildenb** [comment]
@sskeirik can you try out #296 instead? It simplifies the way `#take` and `#drop` work drastically, so this may not be needed.

**Comment by sskeirik** [comment]
I think the other version is much better and should work. Closing unless that proves problematic.

**Pull request closed without merging**